pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -2529,6 +2534,19 @@ public void publishTxnMessage(TxnID txnID, ByteBuf headersAndPayload, PublishCon
             return;
         }
 
+        // need to check transaction buffer ready before before check the message deduplication
+        if (transactionBuffer instanceof TopicTransactionBuffer) {
+            if (!((TopicTransactionBuffer) transactionBuffer).checkIfReady()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}]Transaction buffer not recover complete!", topic);
+                }
+                publishContext.completed(new BrokerServiceException
+                        .ServiceUnitNotReadyException("[{" + topic + "}]Transaction buffer not "
```

Comment:
what about: "Transaction buffer recovery is still running" ?

Reply:
yes